src/view/observer/focusobserver.js

```diff
@@ -22,8 +24,21 @@ export default class FocusObserver extends DomEventObserver {
 		this.domEventType = [ 'focus', 'blur' ];
 	}
 
-	onDomEvent( domEvt ) {
-		this.fire( domEvt.type, domEvt );
+	onDomEvent( domEvent ) {
+		const target = this.document.domConverter.getCorrespondingViewElement( domEvent.target );
+
+		// Update `isFocus` property of root elements.
+		for ( let root of this.document.roots.values() ) {
```

Comment:
I'm not saying that the code would need to be moved to a different class. I'd just add it as a listener in the constructor of `FocusObserver`. That will give us the separation of concerns – that bit fires the event, that uses it.